### PR TITLE
CB-18055 Implement start/StopWithLimitedRetry for Azure

### PR DESCRIPTION
**Pull request by Vatsal23082000**
- start/stopWithLimitedRetry are versions of instance start/stop where it limits the retries that the operation may perform normally.
- If the instances are not started/stopped in the given time than we update the instance state as unknown.

### CODE REVIEW

**Comment by aakashshankar** [comment]
I think a PR was already implemented for this by -> #13177

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:1]: Yes, I have tried it against the recovery flow and it is working perfectly.

[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:1]: You can check it out here:- https://docs.google.com/document/d/19dh5CrcbW-unX7jVGvoyTcPHbazGvCojxwBB1EOiGb0/edit?disco=AAAArPqaR34

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java [hunk 1]
```diff
@@ -291,6 +292,25 @@ public List<CloudVmInstanceStatus> deallocateInstances(AuthenticatedContext ac,
         return statusesAfterDeallocate;
     }
 
+    @Retryable(backoff = @Backoff(delay = 1000, multiplier = 2, maxDelay = 10000), maxAttempts = 5)
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java:295]: Set to a lower value. See the AwsConnector as an example. The intent is to fail fast. In fact, I'm not even sure we want retries here or in AWS ... that can be tracked via a separate PR.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java [hunk 2]
```diff
@@ -291,6 +292,25 @@ public List<CloudVmInstanceStatus> deallocateInstances(AuthenticatedContext ac,
...
+    public List<CloudVmInstanceStatus> deallocateInstances(AuthenticatedContext ac, List<CloudInstance> vms, Long timeboundInMs) {
+        LOGGER.info("Deallocate VMs: {}", vms.stream().map(CloudInstance::getInstanceId).collect(Collectors.toList()));
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java:297]: Log line should ideally change to include the timeout.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java [hunk 3]
```diff
@@ -291,6 +292,25 @@ public List<CloudVmInstanceStatus> deallocateInstances(AuthenticatedContext ac,
...
+        List<CloudVmInstanceStatus> actualVmStatuses = azureVirtualMachineService.getVmsAndVmStatusesFromAzureWithoutRetry(ac, vms).getStatuses();
+        List<CloudVmInstanceStatus> statusesAfterDeallocate = new ArrayList<>();
+        List<CloudInstance> vmsNotStopped = new ArrayList<>();
+        actualVmStatuses.stream().forEach(instance -> {
+            if (InstanceStatus.STOPPED.equals(instance.getStatus())) {
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java:302]: Filter out additional VM instances which may already be in a final state. (See the AwsInstanceConnector as an example).
As an example: If one of the instances is already in a DELETED state - there's no point attempting to STOP it.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java [hunk 4]
```diff
@@ -513,6 +514,13 @@ public Mono<Void> deallocateVirtualMachineAsync(String resourceGroup, String vmN
         return handleException(() -> azure.virtualMachines().deallocateAsync(resourceGroup, vmName));
     }
 
+    public Mono<Void> deallocateVirtualMachineAsync(String resourceGroup, String vmName, Long timeboundInMs) {
+        if (timeboundInMs == null) {
+            return deallocateVirtualMachineAsync(resourceGroup, vmName);
+        }
+        return handleException(() -> azure.virtualMachines().deallocateAsync(resourceGroup, vmName)).timeout(Duration.ofMillis(timeboundInMs));
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java:521]: The timeout here is on the handleException return value.
On the startAsync - it's on the invocation of the startAsync call (as against the handleException)

Which one is it supposed to be ?

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java [hunk 5]
```diff
@@ -320,6 +361,23 @@ private void deallocateInstance(AzureClient azureClient, CloudContext cloudConte
                 .subscribeOn(schedulerProvider.io()));
     }
 
+    private void deallocateInstance(AzureClient azureClient, CloudContext cloudContext, List<CloudVmInstanceStatus> statusesAfterDeallocate,
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java:364]: There's a lot of code duplication between this and deallocateInstances without a timeout .... can the duplicate code be merged into a common method ?

Same comment applies to start, and the various other bits of duplication.
cc @aakashshankar

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java [hunk 6]
```diff
@@ -66,12 +67,45 @@ public List<CloudVmInstanceStatus> start(AuthenticatedContext ac, List<CloudReso
         return statuses;
     }
 
+    @Override
+    public List<CloudVmInstanceStatus> startWithLimitedRetry(AuthenticatedContext ac, List<CloudResource> resources,
+            List<CloudInstance> vms, Long timeboundInMs) {
+        LOGGER.info("Starting vms on Azure: {}", vms.stream().map(CloudInstance::getInstanceId).collect(Collectors.toList()));
```
Comment (sidseth) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:73]: Similarly to stop ... log the timeout value.

Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java]: Changed

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java (hunk 5 above)
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java]: Merged the duplicates

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java (hunk 4 above)
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java]: It should be on the function call.
So, I have updated this

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java (hunk 3 above)
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java]: Done

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java (hunk 2 above)
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java]: Done

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java (hunk 1 above)
Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java]: Done

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java [hunk 7]
```diff
@@ -72,6 +93,13 @@ public List<CloudVmInstanceStatus> stop(AuthenticatedContext ac, List<CloudResou
         return azureUtils.deallocateInstances(ac, vms);
     }
 
+    @Override
+    public List<CloudVmInstanceStatus> stopWithLimitedRetry(AuthenticatedContext ac, List<CloudResource> resources,
+            List<CloudInstance> vms, Long timeboundInMs) {
+        LOGGER.info("Stopping vms on Azure: {}", vms.stream().map(CloudInstance::getInstanceId).collect(Collectors.toList()));
```
Comment (aakashshankar) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:99]: Can we log the timeout here too?

Comment (Vatsal23082000) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java]: Done

**Review by sodre90 — APPROVED** [review]

**Comment by Vatsal23082000** [comment]
Updated the integration test as well.

**Review by lacikaaa — CHANGES_REQUESTED** [review]
please add info to commit message why is this change required, how it would change the behavior

**Review by lacikaaa — APPROVED** [review]
